src/inabox/inabox-resources.js

```diff
@@ -0,0 +1,266 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Deferred} from '../utils/promise';
+import {Observable} from '../observable';
+import {Pass} from '../pass';
+import {Resource, ResourceState} from '../service/resource';
+import {Services} from '../services';
+import {dev} from '../log';
+import {registerServiceBuilderForDoc} from '../service';
+
+const TAG = 'inabox-resources';
+const FOUR_FRAME_DELAY = 70;
+
+/**
+ * @implements {../service/resources-impl.ResourcesDef}
+ */
+class InaboxResources {
+  /**
+   * @param {!../service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @const {!../service/ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @const {!Window} */
+    this.win = ampdoc.win;
+
+    /** @private @const {!Array<!Resource>} */
+    this.resources_ = [];
+
+    /** @private {number} */
+    this.resourceIdCounter_ = 0;
+
+    /** @private @const {!../service/vsync-impl.Vsync} */
+    this.vsync_ = Services./*OK*/ vsyncFor(this.win);
+
+    /** @const @private {!Pass} */
+    this.pass_ = new Pass(this.win, this.doPass_.bind(this));
+
+    /** @private @const {!Observable} */
+    this.passObservable_ = new Observable();
+
+    /** @const @private {!Deferred} */
+    this.firstPassDone_ = new Deferred();
+  }
+
+  /** @override */
+  get() {
+    return this.resources_.slice(0);
+  }
+
+  /** @override */
+  getAmpdoc() {
+    return this.ampdoc_;
+  }
+
+  /** @override */
+  getMeasuredResources(unusedHostWin, unusedFilterFn) {
```

Comment:
This is a bit dangerous. Some code might rely on width/height being known which this would not provide for. Also there might be a key question of non-displayable resources behavior.